MAINT: add test to prevent new public-looking modules being added

rgommers opened the pull request:
There is more work to do (added a FIXME), but that's for another time. Preventing to add new submodules is a good start. Also added some documentation, and split things into public and private submodules with comments.

The change to `numpy.lib.mixins.__all__` shows that people have been working around our `import *` mess sometimes. Not adding some public object to an `__all__` dict is incorrect of course - need a more structural fix by getting rid of `*` imports.

rgommers committed:
TST: add test to prevent new public-looking modules being added

rgommers commented:
Hmm, we start with a failure that I can't reproduce locally:
```
>           raise AssertionError("Found unexpected modules: {}".format(modnames))
E           AssertionError: Found unexpected modules: ['numpy.core.generate_numpy_api']
```
shows up for an install, but not an in-place build. adding it to the list

rgommers committed:
TST: add `core/generate_numpy_api.py` to private-but-present modules

rgommers committed:
MAINT: fix issue with `np.lib.mixins.__all__` and import of mixins

rgommers committed:
TST: add details to the public modules test about how to check public-ness

rgommers committed:
MAINT: add missing 'Arrayterator' to `numpy.lib.__all__`

Also finish the TODO about figuring out which np.lib.<submodule>'s
are public.

This is a giant mess ...

eric-wieser commented:
> The change to numpy.lib.mixins.__all__ shows that people have been working around our import * mess sometimes.

Can you elaborate on what you mean by this?

eric-wieser commented on numpy/tests/test_public_api.py:
```diff
@@ -109,3 +117,299 @@ def test_NPY_NO_EXPORT():
+        return False
+
+    if name.startswith("numpy."):
+        name = name[6:]
```
Is there a case when you expect it not to start with `numpy.`? And if that does happen, doesn't it then result in incorrect lookups in your name lists?

rgommers replied:
nope, all starts with `numpy.`. change to have full names in `PUBLIC_MODULES` et al.

eric-wieser commented on numpy/tests/test_public_api.py:
```diff
@@ -109,3 +117,299 @@ def test_NPY_NO_EXPORT():
+        for objname in objnames:
+            if not objname.startswith('_'):
+                fullobjname = modname + '.' + objname
+                if isinstance(eval(fullobjname), types.ModuleType):
```
Why not just `getattr(module, objname)`?

rgommers replied:
thanks, done. was leftover from an older version where I didn't have `module`

eric-wieser commented on numpy/tests/test_public_api.py:
```diff
@@ -109,3 +117,299 @@ def test_NPY_NO_EXPORT():
+                print(obj)
+
+    """
+    modnames = []
```
nit: I think it would be clearer to make this local to `check`, and change the function signature to `members = find_unexpected_members(mod_name)`

rgommers replied:
done

eric-wieser commented on numpy/tests/test_public_api.py:
```diff
@@ -109,3 +117,299 @@ def test_NPY_NO_EXPORT():
+]
+
+
+PRIVATE_BUT_PRESENT_MODULES = [
```
A summary comment would be good for this and `PUBLIC_ALIASED_MODULES`

rgommers replied:
done

rgommers committed:
DOC: add some comments to explain namespace cleanup in numpy/__init__.py

[ci skip]

rgommers commented:
> > The change to numpy.lib.mixins.**all** shows that people have been working around our import * mess sometimes.
> 
> Can you elaborate on what you mean by this?

The comment explicitly said not to add `NDArrayOperatorsMixin` because it would then end up in the main `numpy` namespace. While that goal is good, it's incorrect to have a public function not in the `__all__` dict in the file it's in. Also, this effectively has made `numpy.lib.mixins` public, because the mixin didn't end up in `numpy.lib` even though there's a `from .mixins import *` in `numpy/lib/__init__.py`; I'm not sure what the intent was there.

The root cause is that in our main `__init__.py` there's a `from .lib import *`, which makes little sense.

rgommers committed:
TST: add comment on public/semi-private modules.

Also address other review comments.

rgommers committed:
TST: make xfailed test for modules in public API pass.

mattip commented on numpy/tests/test_public_api.py:
```diff
@@ -109,3 +117,341 @@ def test_NPY_NO_EXPORT():
+    "random.bit_generator",
+    "random.bounded_integers",
+    "random.common",
+    "random.entropy",
```
This conflicts with gh-14498, whichever is merged first then the other will have to remove this line.

mattip replied:
Also: `random.info` does not exist. Do the tests check that all these actually are present?

rgommers replied:
No, they were generated from an install and the checks are for no additions, not for removals (that'd be either a good thing or caused by tests, so less interesting to test from my perspective).

rgommers replied:
> Also: `random.info` does not exist.

this is not true, there's a `numpy/random/info.py` in current master

rgommers replied:
removed `random.entropy`

seberg replied:
I think it would be preferable if the tests check whether they are exactly present (unless there is some reason that is unreliable). It is good to know when we remove it, and when we do, we want to make sure we do not accidentally readd.

rgommers replied:
Okay done. That did identify one more stray file to clean up, and will help keep things clean.

Possible this isn't 100% robust, since importing may fail on different platforms with different exceptions - let's deal with that if/when such cases show up.

rgommers replied:
CI all green on four platforms without anything unexpected, so robustness probably okay. comments here addressed, so resolving this one.

rgommers committed:
TST: remove random.entropy from public API test, it was removed in gh-14498

rgommers committed:
TST: add test to check that (semi-)public API modules can be imported

Also remove `numpy.ma.version.py`, it was not importable and served
no purpose.

mattip commented:
Thanks @rgommers. I guess the next step is to identify candidates for namespace deprecation.

rgommers commented:
>  I guess the next step is to identify candidates for namespace deprecation.

Before that, lower-hanging fruit may be to remove `info.py` files, add some `__all__` dicts, and clean up existing namespaces (e.g. removing `numpy.math` really shouldn't require a deprecation).

charris commented on numpy/tests/test_public_api.py:
```diff
@@ -109,3 +118,390 @@ def test_NPY_NO_EXPORT():
+    "random.philox",
+    "random.sfc64",
+    "testing.decorators",
+    "testing.noseclasses",
```
This broke the wheel builds, I suspect because nose is not present in those builds. Nose doesn't support recent Python3 out of the box.

rgommers replied:
do you have a link?

charris replied:
`nose` is optional, and if not present an import error will be raised if a attempt to import it is made.

charris replied:
https://travis-ci.org/MacPython/numpy-wheels/builds/587676702?utm_medium=notification&utm_source=email

rgommers replied:
thanks. okay that's easy to fix, PR to follow
